// @engine/table/bin/modules/cell/cell_controllers/cell_controller.crud.js
import { requestGet, requestPost } from '@engine/crud';
/**
 * 获取数据字典的基础方法
 * @param widgetName 控件名称
 * @param op 其余参数
 */
function fetchCellBasicControlAttr(widgetName, op = {}) {
    return requestPost('cell/widget/data', {
        data: Object.assign(Object.assign({}, op), { widgetname: widgetName }),
    });
}
/**
 * 下拉框控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchCellComboControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, op);
}
/**
 * 请求cell view value的最基础请求
 * @param params params
 */
export function fetchCellWidgetViewValue(params) {
    return requestGet('cell/widget/view/value', {
        params,
    });
}
/**
 * 日期控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchCellDatetimeControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, op);
}
/**
 * 按钮组控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchCellGroupControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, op);
}
/**
 * 老树控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchOldCellTreeControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, op);
}
/**
 * 新树控件获取单个节点下的树形结构
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchNewCellTreeStructure(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, Object.assign(Object.assign({}, op), { fine_outlet_widget: 'nx', completion: true }));
}
/**
 * 获取新树控件返回值
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchNewCellTreeControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, Object.assign(Object.assign({}, op), { fine_outlet_widget: 'nx' }));
}
/**
 * 网页框控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchIframeControlAttr(widgetName, op = {}) {
    return fetchCellBasicControlAttr(widgetName, op);
}
