// @engine/controller/bin/modules/tag_combo_checkbox/multiselectcombo/multiselectcombo.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var ParameterMultiSelectCombo_1;
import { ParameterMultiSelectTrigger } from './multiselecttrigger';
import { shortcut } from '@engine/core';
import { ParameterMultiSelectPopup } from './multiselectpopup';
import { PopupMaxHeight } from '../../controller.service';
import { ParameterMultiSelectSwitcher } from './multiselectswitcher';
let ParameterMultiSelectCombo = ParameterMultiSelectCombo_1 = class ParameterMultiSelectCombo extends BI.Widget {
    constructor() {
        super(...arguments);
        this.repopulate = false;
        this.initialized = false;
        this.needResize = false;
        this.props = {
            baseCls: 'bi-search-text-value-combo',
            height: 24,
            itemHeightGetter: () => 24,
            text: '',
            items: [],
            tipType: '',
            warningTitle: '',
            attributes: {
                tabIndex: 0,
            },
            title: '',
            value: [],
            itemsCreator: BI.emptyFn,
            watermark: '',
            allowEdit: false,
            customData: true,
            allowCount: false,
        };
    }
    render() {
        const { height, value, text, tipType, warningTitle, title, itemsCreator, watermark, allowEdit, customData, itemHeightGetter, allowCount } = this.options;
        return {
            type: BI.AbsoluteLayout.xtype,
            scrollable: false,
            items: [{
                    el: {
                        type: BI.Combo.xtype,
                        adjustLength: 2,
                        toggle: !allowEdit,
                        ref: (_ref) => {
                            this.combo = _ref;
                        },
                        el: {
                            type: ParameterMultiSelectTrigger.xtype,
                            customData,
                            watermark,
                            allowEdit,
                            cls: 'search-text-value-trigger bi-border-radius',
                            ref: (_ref) => {
                                this.trigger = _ref;
                            },
                            height: height - 2,
                            text,
                            value,
                            tipType,
                            warningTitle,
                            itemHeightGetter,
                            itemsCreator,
                            valueGetter: () => this.popup ? this.popup.getValue() : value,
                            unmatchedValueGetter: () => this.popup ? this.popup.getUnmatchedValues() : value,
                            title,
                            listeners: [{
                                    eventName: ParameterMultiSelectTrigger.EVENT_CONFIRM,
                                    action: () => {
                                        this.combo.hideView();
                                        // this.fireEvent(ParameterMultiSelectCombo.EVENT_CHANGE);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_FOCUS,
                                    action: () => {
                                        this.showView();
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_BEFORE_EDIT);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_CHANGE,
                                    action: (text, value, isSelected) => {
                                        this.popup.switchState(text, value, isSelected);
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_CLICK_ITEM);
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_AFTER_EDIT);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_SELECT_ALL,
                                    action: (isSelected, items) => {
                                        this.popup.switchStates(items, isSelected);
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_AFTER_EDIT);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_CLEAR,
                                    action: () => {
                                        this.reset();
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_SEARCHING,
                                    action: () => {
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_AFTER_EDIT);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_BLUR,
                                    action: () => {
                                        this.fireEvent(ParameterMultiSelectCombo_1.EVENT_BLUR);
                                    },
                                }, {
                                    eventName: ParameterMultiSelectTrigger.EVENT_CLICK,
                                    action: () => {
                                        this.showView();
                                    },
                                }],
                        },
                        popup: {
                            minWidth: 135,
                            maxHeight: PopupMaxHeight,
                            el: {
                                type: ParameterMultiSelectPopup.xtype,
                                itemHeightGetter,
                                itemsCreator,
                                ref: (_ref) => {
                                    this.popup = _ref;
                                    this.trigger.setAdapter(this.popup);
                                },
                                repopulateGetter: () => this.repopulate,
                                listeners: [{
                                        eventName: ParameterMultiSelectPopup.EVENT_CHANGE,
                                        action: () => {
                                            this.setStateValue(this.popup.getValue());
                                            this.fireEvent(ParameterMultiSelectCombo_1.EVENT_CLICK_ITEM);
                                            this.fireEvent(ParameterMultiSelectCombo_1.EVENT_AFTER_EDIT);
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_SELECT_ALL,
                                        action: () => {
                                            this.setValue(this.popup.getValue());
                                            this.fireEvent(ParameterMultiSelectCombo_1.EVENT_AFTER_EDIT);
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_CONFIRM,
                                        action: () => {
                                            this.combo.hideView();
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_CLEAR,
                                        action: () => {
                                            this.reset();
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_RESIZE,
                                        action: () => {
                                            this.numberCounter.initSwitcher();
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_SWITCH_STATE,
                                        action: () => {
                                            this.numberCounter.updateValue();
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_LOADED,
                                        action: () => {
                                            this.initialized = true;
                                            BI.nextTick(() => {
                                                // 视情况更新搜索结果弹窗高度，如果搜索请求先于获取全部字典调用，就需要手动调整
                                                this.trigger.isSearching() && this.trigger.adjustPopupHeight(this.popup);
                                            });
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectPopup.EVENT_LOADING,
                                        action: () => {
                                            this.initialized = false;
                                        },
                                    }],
                            },
                            value,
                        },
                        listeners: [{
                                eventName: BI.Combo.EVENT_AFTER_HIDEVIEW,
                                action: () => {
                                    const keyword = this.trigger.isSearching() ? this.getAdapter().getKeyword() : undefined;
                                    this.trigger.stopEditing();
                                    this.numberCounter.hideView();
                                    this.popup && this.setStateValue(this.popup.getValue());
                                    this.fireEvent(ParameterMultiSelectCombo_1.EVENT_CONFIRM, keyword);
                                },
                            }, {
                                eventName: BI.Combo.EVENT_BEFORE_POPUPVIEW,
                                action: () => {
                                    this.fireEvent(ParameterMultiSelectCombo_1.EVENT_BEFORE_POPUPVIEW);
                                    if (this.repopulate) {
                                        this.populate([]);
                                        this.repopulate = false;
                                    }
                                },
                            }, {
                                eventName: BI.Combo.EVENT_AFTER_POPUPVIEW,
                                action: () => {
                                    this.checkResize();
                                },
                            }],
                        hideChecker: (e) => this.triggerBtn.element.find(e.target).length === 0 && this.numberCounter.element.find(e.target).length === 0,
                    },
                    left: 0,
                    right: 0,
                    bottom: 0,
                    top: 0,
                }, {
                    el: {
                        type: BI.TriggerIconButton.xtype,
                        cls: 'trigger-icon-button',
                        ref: (_ref) => {
                            this.triggerBtn = _ref;
                        },
                        width: height,
                        height,
                        handler: () => {
                            this.numberCounter.hideView();
                            if (this.combo.isViewVisible()) {
                                this.combo.hideView();
                                return;
                            }
                            this.showView();
                            this.fireEvent(ParameterMultiSelectCombo_1.EVENT_BEFORE_EDIT);
                        },
                    },
                    right: 0,
                    bottom: 0,
                    top: 0,
                },
                {
                    el: {
                        type: BI.VerticalAdaptLayout.xtype,
                        invisible: !allowCount,
                        items: [{
                                type: ParameterMultiSelectSwitcher.xtype,
                                valueGetter: () => this.popup ? this.popup.getValue() : this.options.value,
                                popupGetter: () => this.popup,
                                itemHeightGetter,
                                ref: (_ref) => {
                                    this.numberCounter = _ref;
                                },
                                listeners: [{
                                        eventName: ParameterMultiSelectSwitcher.EVENT_CHANGE,
                                        action: () => {
                                            if (!this.combo.isViewVisible()) {
                                                this.combo.showView();
                                            }
                                        },
                                    }, {
                                        eventName: ParameterMultiSelectSwitcher.EVENT_BEFORE_POPUPVIEW,
                                        action: () => {
                                            this.numberCounter.updateValue();
                                        },
                                    }, {
                                        eventName: BI.Events.VIEW,
                                        action: (visbility) => {
                                            BI.nextTick(() => {
                                                if (visbility) {
                                                    this.trigger.refreshPlaceHolderWidth(this.numberCounter.element.outerWidth() * 1.5);
                                                }
                                            });
                                        },
                                    }],
                            }],
                    },
                    height,
                    top: 0,
                    right: 24,
                }],
        };
    }
    reset() {
        this.popup && this.popup.setAllUnselected();
        this.setValue([]);
        this.popup && this.popup.checkAllSelected();
    }
    showView() {
        this.combo.showView();
    }
    checkResize() {
        if (this.popup && this.needResize) {
            this.popup.resize(true);
            this.needResize = false;
        }
    }
    populate(items) {
        this.options.items = items;
        this.combo.populate();
    }
    // 仅set状态（trigeer + numberCounter）
    setStateValue(v = []) {
        this.options.value = v;
        this.trigger.setValue(v);
        this.numberCounter.setValue(v);
    }
    setValue(v = []) {
        this.options.value = v;
        this.combo.setValue(v);
        this.numberCounter.setValue(v);
    }
    getValue() {
        return this.popup ? this.popup.getValue() : this.options.value;
    }
    setRepopulate(v) {
        this.reset();
        this.popup && this.popup.resetState();
        // 如果初始化过，则需要重新初始化
        if (this.popup && this.initialized) {
            this.combo.isViewVisible() ? this.combo.populate() : this.repopulate = v;
            this.needResize = true;
        }
    }
    focus() {
        this.trigger.focus();
        this.combo.showView();
        // TODO: 需要研究下为啥事件没触发
        this.checkResize();
    }
    blur() {
        this.trigger.blur();
        this.combo.hideView();
    }
    getAllButtons() {
        return this.popup.getAllButtons();
    }
    getAdapter() {
        return this.trigger.getSearcher();
    }
    setText(v) {
        this.trigger.setText(v);
    }
    addValue(text, value) {
        this.popup.switchState(text, value);
        if (!this.combo.isViewVisible()) {
            this.needResize = true;
        }
    }
    setWaterMark(v) {
        this.trigger.setWaterMark(v);
    }
};
ParameterMultiSelectCombo.EVENT_CHANGE = 'EVENT_CHANGE';
ParameterMultiSelectCombo.EVENT_BEFORE_POPUPVIEW = 'EVENT_BEFORE_POPUPVIEW';
ParameterMultiSelectCombo.EVENT_CONFIRM = 'EVENT_CONFIRM';
ParameterMultiSelectCombo.xtype = 'report.search_text_value_combo';
ParameterMultiSelectCombo.EVENT_BEFORE_EDIT = 'EVENT_BEFORE_EDIT';
ParameterMultiSelectCombo.EVENT_AFTER_EDIT = 'EVENT_AFTER_EDIT';
ParameterMultiSelectCombo.EVENT_CLICK_ITEM = 'EVENT_CLICK_ITEM';
ParameterMultiSelectCombo.EVENT_BLUR = 'EVENT_BLUR';
ParameterMultiSelectCombo = ParameterMultiSelectCombo_1 = __decorate([
    shortcut()
], ParameterMultiSelectCombo);
export { ParameterMultiSelectCombo };
