// @engine/controller/bin/modules/password/password.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { WIDTH, HEIGHT } from '../controller.constants';
import { setEditorStyle, setFontSize } from '../controller.service';
import { validateParamPassword, getParamPasswordErrorMsg } from './password.service';
import { shortcut } from '@engine/core';
import { AbstractParameterControllerWidget } from '../controller.abstract';
let ParameterPassword = class ParameterPassword extends AbstractParameterControllerWidget {
    constructor() {
        super(...arguments);
        this.props = {
            baseCls: 'report-main-parameter-container-controller-password',
            $testId: 'report-main-parameter-container-controller-password',
            width: WIDTH,
            height: HEIGHT,
            value: '',
            passwordText: true,
            disabled: false,
            invisible: false,
            needSubmit: false,
            fontSize: 12,
            allowBlank: true,
            errorMsg: BI.i18nText('Fine-Report_Preview_Can_Not_Null'),
            watermark: '',
            regex: '',
            regErrorMsg: BI.i18nText('Fine-Report_Preview_Input_Error'),
            allowBg: true,
        };
    }
    render() {
        const { width, height, value, fontSize, watermark, disabled, allowBg } = this.options;
        const editorCls = setFontSize('report-main-parameter-container-controller-password', fontSize);
        return {
            type: BI.VerticalLayout.xtype,
            cls: allowBg ? 'bi-card' : '',
            items: [{
                    type: BI.Editor.xtype,
                    cls: `${editorCls} bi-border bi-border-radius bi-focus-shadow`,
                    ref: (_ref) => {
                        this.editor = _ref;
                    },
                    width: width - 2,
                    height: height - 2,
                    disabled,
                    allowBlank: true,
                    watermark,
                    value,
                    inputType: 'password',
                    listeners: [{
                            eventName: BI.TextEditor.EVENT_FOCUS,
                            action: () => {
                                this.hideBubble();
                                !this.silence && this.fireEvent('beforeedit');
                            },
                        }, {
                            eventName: BI.TextEditor.EVENT_KEY_DOWN,
                            action: () => {
                                const result = !this.silence && this.fireEvent('afteredit');
                                result !== false && this.fireEvent('EVENT_CHANGE', this.editor.getValue());
                            },
                        }, {
                            eventName: BI.TextEditor.EVENT_BLUR,
                            action: () => {
                                !this.silence && this.fireEvent('stopedit');
                            },
                        }],
                }],
        };
    }
    // 公开接口
    setWaterMark(v) {
        this.editor.setWaterMark(v);
    }
    async setControllerValue(v) {
        this.editor.setValue(v);
    }
    setInteractValue(v) {
        this.setControllerValue(v);
    }
    getBasicValidation(value = this.editor.getValue()) {
        const v = BI.isArray(value) ? value.map(v => `${v}`).join('') : value;
        const { allowBlank, regex, errorMsg, regErrorMsg } = this.options;
        return validateParamPassword(v, { allowBlank, regex }) ? {
            result: true,
        } : {
            result: false,
            errorMsg: getParamPasswordErrorMsg(v, { allowBlank, errorMsg, regErrorMsg }),
        };
    }
    preGetValidation() {
        const { value } = this.options;
        return this.getBasicValidation(value);
    }
    postGetValidation() {
        return this.getBasicValidation();
    }
    postValidateAndToast() {
        const validation = this.getValidation();
        if (validation.result) {
            return true;
        }
        this.showBubble(validation.errorMsg);
        return false;
    }
    emitManualStopEditEvent() {
        this.fireEvent('stopedit');
    }
    getValue() {
        return this.editor.getValue();
    }
    reset() {
        this.editor.setValue();
    }
    resize(width, height) {
        const restoreValue = this.editor.getValue();
        super.resize(width, height);
        this.editor.setValue(restoreValue);
    }
    getAdapter() {
        return this.editor;
    }
    focus() {
        this.editor.focus();
    }
    blur() {
        this.editor.blur();
    }
    setStyle(styles) {
        this.editor.element.addClass(setEditorStyle(styles));
    }
    getText() {
        return this.editor.getValue();
    }
};
ParameterPassword.xtype = 'report.main.parameter_container.controller.password';
ParameterPassword = __decorate([
    shortcut()
], ParameterPassword);
export { ParameterPassword };
