// @engine/controller/bin/modules/controller.crud.js
import { envProvider, universalEnvProvider } from '@engine/core';
import { requestGet, requestPost } from '@engine/crud';
/**
 * 获取数据字典的基础方法
 * @param widgetName 控件名称
 * @param op 其余参数
 */
function fetchParamBasicControlAttr(widgetName, op = {}, cancelToken) {
    return requestPost('widget/data', {
        data: Object.assign(Object.assign({}, op), { widgetname: widgetName }),
        baseURL: universalEnvProvider.ReqPrefix,
        cancelToken,
    }).catch(() => { });
}
/**
 * 下拉框控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchParamControlAttr(widgetName, op = {}) {
    return fetchParamBasicControlAttr(widgetName, op);
}
/**
 * 请求view value的最基础请求
 * @param data data
 */
export async function fetchWidgetViewValue(data) {
    return requestPost('widget/view/value', {
        data,
        baseURL: universalEnvProvider.ReqPrefix,
    });
}
/**
 * 日期控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchParamDatetimeControlAttr(widgetName, op = {}) {
    return fetchParamBasicControlAttr(widgetName, op);
}
/**
 * 老树控件获取数据字典
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchOldParamTreeControlAttr(widgetName, op = {}, cancelToken) {
    return fetchParamBasicControlAttr(widgetName, op, cancelToken).then(v => v || []);
}
/**
 * 新树控件获取单个节点下的树形结构
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchNewParamTreeStructure(widgetName, op = {}) {
    return fetchParamBasicControlAttr(widgetName, Object.assign(Object.assign({}, op), { fine_outlet_widget: 'nx', completion: true }));
}
/**
 * 获取新树控件返回值
 * @param widgetName 控件名称
 * @param op 其余参数
 */
export function fetchNewParamTreeControlAttr(widgetName, op = {}, cancelToken) {
    return fetchParamBasicControlAttr(widgetName, Object.assign(Object.assign({}, op), { fine_outlet_widget: 'nx' }), cancelToken);
}
/**
 * 获取下载文件地址
 * @param fileId
 */
export function getDownloadFileUrl(fileId) {
    return `${universalEnvProvider.ReqPrefix}/attach/download?id=${fileId}&sessionID=${envProvider.SessionID}`;
}
/**
 * 下载文件
 * @param fileId
 */
export function downloadFile(fileId) {
    window.open(getDownloadFileUrl(fileId));
}
/**
 * 删除上传的文件
 * @param fileId
 */
export function deleteFile(fileId) {
    return requestGet('attach/release', {
        params: {
            id: fileId,
        },
        baseURL: universalEnvProvider.ReqPrefix,
    });
}
