// @engine/controller/bin/modules/combo/singleselectcombo/singleselectsearchpopup.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { shortcut } from '@engine/core';
import { ParameterSingleSelectPopup } from './singleselectpopup';
import { COMBO_POPUP_HEIGHT } from './singleselectcombo.constants';
let ParameterSingleSelectSearchPopup = class ParameterSingleSelectSearchPopup extends ParameterSingleSelectPopup {
    constructor() {
        super(...arguments);
        this.props = {
            itemsCreator: BI.emptyFn,
            keywordGetter: () => undefined,
            valueGetter: () => undefined,
            unmatchedValueGetter: () => [],
            minWidth: 135,
            maxHeight: 400,
            height: undefined,
            value: undefined,
            customData: true,
            allowNoSelect: false,
            repopulateGetter: () => false,
            name4EmptyGetter: () => '',
            itemHeightGetter: () => 24,
            maxCountGetter: () => 0,
            popupAdjuster: () => { },
            tipText: BI.i18nText('BI-No_Selected_Item'),
        };
    }
    transformItems(items) {
        const { valueGetter, unmatchedValueGetter, keywordGetter, customData } = this.options;
        const keyword = keywordGetter() || '';
        const unmatchedValues = unmatchedValueGetter().slice(0);
        let selectedValue = valueGetter();
        items.forEach(item => {
            let isMatched = false;
            if (BI.isNull(selectedValue)) {
                isMatched = true;
            }
            else {
                isMatched = selectedValue.text === item.text && selectedValue.value === item.value;
                item.selected = isMatched;
                if (isMatched) {
                    selectedValue = undefined;
                }
            }
            const unmatchedValueIndex = unmatchedValues.findIndex(value => value.text === item.text && value.value === item.value);
            if (unmatchedValueIndex !== -1) {
                unmatchedValues.splice(unmatchedValueIndex, 1);
            }
        });
        unmatchedValues.forEach(item => {
            let isMatched = false;
            if (BI.isNull(selectedValue)) {
                isMatched = true;
            }
            else {
                isMatched = selectedValue.text === item.text && selectedValue.value === item.value;
                item.selected = isMatched;
                if (isMatched) {
                    selectedValue = undefined;
                }
            }
        });
        // 剩下的值如果能匹配的肯定不是在数据字典中的，也就是text和item相同的项
        const { match, find } = BI.Func.getSearchResult(unmatchedValues, keyword, 'value');
        const allItems = find.concat(items, match);
        if (BI.size(allItems) === 0 && customData) {
            this.addTextBtn.setText(BI.i18nText('BI-Basic_Click_To_Add_Text', keywordGetter()));
            this.addTextBtn.visible();
        }
        else {
            this.addTextBtn.invisible();
        }
        return match.concat(items, find);
    }
    resizeLayout(items) {
        this.vtapeWrapper.attr('items')[0].height = BI.size(items) !== 0 ? 0 : COMBO_POPUP_HEIGHT;
        this.vtapeWrapper.resize();
    }
};
ParameterSingleSelectSearchPopup.xtype = 'report-single-select-search-popup';
ParameterSingleSelectSearchPopup = __decorate([
    shortcut()
], ParameterSingleSelectSearchPopup);
export { ParameterSingleSelectSearchPopup };
