// @engine/core/bin/modules/core.constants.js
export const LOADED = 'LOADED';
export const LOADING = 'LOADING';
export const LOADERROR = 'LOADERROR';
export const NOTLOADED = 'NOTLOADED';
export const ICON_BUTTON_TYPES = {
    FIRST_PAGE: 'first',
    PRE_PAGE: 'previous',
    CUR_PAGE: 'page-navi',
    NEXT_PAGE: 'next',
    LAST_PAGE: 'last',
    SCALE_ADJUST: 'scale',
    MAIL: 'mail',
    EXPORT: 'excel-menu',
    PDF: 'export-pdf',
    PAGE_EXCEL: 'export-excel-page',
    ORIGIN_EXCEL: 'export-excel-ori',
    PAGE_SHEET_EXCEL: 'export-excel-sheet',
    WORD: 'export-word',
    PRINT: 'print',
    CUSTOM_BUTTON: 'button',
    COMPATIBLE_PRINT: 'compatible-print',
    FLASH_PRINT: 'flash-print',
    APPLET_PRINT: 'applet-print',
    PDF_PRINT: 'pdf-print',
    OFFSET_PRINT: 'offset-print',
    SUBMIT: 'submit',
    TEMP_SAVE: 'temp-save',
    CHECK_DATA: 'check-data',
    CLEAR_DATA: 'clear-data',
    INSERT_DATA: 'insert-data',
    DELETE_DATA: 'delete-data',
    // TODO: 后续要和后台沟通换掉这个键
    SHOW_CELL_VALUE: 'label',
    IMPORT_EXCEL: 'import-excel',
    CUSTOM_IMPOERT_EXCEL: 'custom-import-excel',
    PRINT_PREVIEW: 'print-preview',
    IMAGE_EXPORT: 'export-image',
    PAGE_SETTING: 'page-setting',
    MENU_BUTTON: 'menubutton',
};
export const ICON_FONT_MAP = new Map([
    [ICON_BUTTON_TYPES.SUBMIT, 'submit-font'],
    ['submitall', 'submit-font'],
    ['pdf', 'export-pdf-font'],
    ['excel', 'export-excel-font'],
    ['word', 'export-word-font'],
    ['image', 'export-image-font'],
    ['appletprint', 'print-applet-font'],
    [ICON_BUTTON_TYPES.LAST_PAGE, 'last-page-font'],
    [ICON_BUTTON_TYPES.NEXT_PAGE, 'next-page-font'],
    [ICON_BUTTON_TYPES.FIRST_PAGE, 'first-page-font'],
    [ICON_BUTTON_TYPES.PRE_PAGE, 'pre-page-font'],
    ['pdfprint', 'print-pdf-font'],
    ['flashprint', 'print-flash-font'],
    ['customizexcel', 'import-excel-font'],
    ['stash', 'temp-save-font'],
    [ICON_BUTTON_TYPES.PRINT, 'print-font'],
    ['export', 'export-font'],
    ['setprinteroffset', 'print-offset-font'],
    ['email', 'mail-font'],
    ['appendrow', 'add-record-font'],
    ['deleterow', 'delete-record-font'],
    ['preview', 'print-preview-font'],
    ['pagesetup', 'page-setting-font'],
    ['frozzen', 'show-cell-font'],
    ['clearstash', 'clear-data-font'],
    ['verify', 'check-data-font'],
]);
