// @engine/table/bin/modules/cell/cell_controllers/tree/cell_controller.tree.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { HEIGHT, ParameterTree, WIDTH } from '@engine/controller';
import { shortcut } from '@engine/core';
import { AbstractParameterCellControllerWidget } from '../cell_controller.abstract';
import { fetchNewCellTreeControlAttr, fetchNewCellTreeStructure, fetchOldCellTreeControlAttr } from '../cell_controller.crud';
let CellTree = class CellTree extends AbstractParameterCellControllerWidget {
    constructor() {
        super(...arguments);
        this.props = {
            width: WIDTH,
            height: HEIGHT,
            fontSize: 12,
            reportIndex: 0,
            position: {
                row: 0,
                column: 0,
            },
            xlisteners: [],
            baseCls: '',
        };
    }
    render() {
        const _a = this.options, { xlisteners, position, reportIndex, baseCls } = _a, opts = __rest(_a, ["xlisteners", "position", "reportIndex", "baseCls"]);
        return Object.assign(Object.assign({}, opts), { fontSize: 12, type: ParameterTree.xtype, ref: (ref) => {
                this.tree = ref;
            }, listeners: xlisteners, oldTreeControlAttrFetcher: fetchOldCellTreeControlAttr, newTreeControlAttrFetcher: fetchNewCellTreeControlAttr, newTreeStructureFetcher: fetchNewCellTreeStructure, extraOptsGetter: () => {
                return {
                    location: Object.assign(Object.assign({}, position), { reportIndex }),
                };
            }, beforeRender(cb) {
                this.updateControllAttr()
                    .then(() => cb());
            } });
    }
    // 复写父类不触发afterinit事件
    mounted() { }
    getValue() {
        return this.tree.getValue();
    }
    postGetValidation() {
        return this.tree.postGetValidation();
    }
    preGetValidation() {
        return this.tree.preGetValidation();
    }
    async setControllerValue(v, data) {
        await this.tree.setControllerValue(v, data);
    }
    reset() {
        this.tree.reset();
    }
    postValidateAndToast() {
        return this.tree.postValidateAndToast();
    }
    setInteractValue(v) {
        this.tree.setInteractValue(v);
    }
    reloadControlAttr() {
        this.tree.reloadControlAttr();
    }
    setStyle() { }
};
CellTree.xtype = 'report.main.table_cell.controller.tree';
CellTree = __decorate([
    shortcut()
], CellTree);
export { CellTree };
