// @engine/controller/bin/modules/number/number.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { WIDTH, HEIGHT } from '../controller.constants';
import { setEditorStyle, setFontSize, setLineHeight } from '../controller.service';
import { checkValidation, transformValidation2ErrorMsg } from './number.service';
import { shortcut } from '@engine/core';
import { AbstractParameterControllerWidget } from '../controller.abstract';
let ParameterNumber = class ParameterNumber extends AbstractParameterControllerWidget {
    constructor() {
        super(...arguments);
        this.props = {
            baseCls: 'report-main-parameter-container-controller-number',
            $testId: 'report-main-parameter-container-controller-number',
            width: WIDTH,
            height: HEIGHT,
            value: undefined,
            allowDecimals: true,
            allowNegative: true,
            maxDecLength: 16,
            maxIntLength: 32,
            maxValue: undefined,
            minValue: undefined,
            regErrorMsg: '',
            disabled: false,
            invisible: false,
            needSubmit: false,
            fontSize: 12,
            allowBlank: true,
            errorMsg: BI.i18nText('Fine-Report_Preview_Can_Not_Null'),
            watermark: '',
            allowBg: true,
        };
    }
    render() {
        const { width, height, value, fontSize, watermark, disabled, allowBg } = this.options;
        const editorCls = setFontSize('report-main-parameter-container-controller-number', fontSize);
        const editorCls2 = setLineHeight('report-main-parameter-container-controller-number', height);
        return {
            type: BI.VerticalLayout.xtype,
            cls: allowBg ? 'bi-card' : '',
            items: [{
                    type: BI.TextEditor.xtype,
                    cls: `${editorCls} ${editorCls2} bi-border-radius`,
                    ref: (_ref) => {
                        this.editor = _ref;
                    },
                    vgap: 0,
                    width,
                    height,
                    watermark,
                    allowBlank: true,
                    disabled,
                    // REPORT-19456 特殊交互，数值控件如果默认值非法不显示值
                    // https://kms.fineres.com/pages/viewpage.action?pageId=253971984 不再进行初始化时的校验
                    value,
                    listeners: [{
                            eventName: BI.TextEditor.EVENT_FOCUS,
                            action: () => {
                                this.hideBubble();
                                !this.silence && this.fireEvent('beforeedit');
                            },
                        }, {
                            eventName: BI.TextEditor.EVENT_KEY_DOWN,
                            action: () => {
                                this.hideBubble();
                                const result = !this.silence && this.fireEvent('afteredit');
                                result !== false && this.fireEvent('EVENT_CHANGE', this.editor.getValue());
                            },
                        }, {
                            eventName: BI.TextEditor.EVENT_BLUR,
                            action: () => {
                                !this.silence && this.fireEvent('stopedit');
                            },
                        }],
                }],
        };
    }
    async setControllerValue(v) {
        const validation = this.getBasicValidation(v);
        validation.result && this.editor.setValue(v);
    }
    getValue() {
        return this.editor.getValue();
    }
    getBasicValidation(value) {
        var _a;
        if (value === void 0) { value = (_a = this.editor) === null || _a === void 0 ? void 0 : _a.getValue(); }
        const { allowNegative, allowDecimals, allowBlank, maxDecLength, maxIntLength, maxValue, minValue, errorMsg, regErrorMsg } = this.options;
        const validation = checkValidation(value, { allowNegative, allowDecimals, allowBlank, maxDecLength, maxIntLength, maxValue, minValue });
        return validation === true ? {
            result: true,
        } : {
            result: false,
            errorMsg: transformValidation2ErrorMsg(validation, {
                errorMsg,
                regErrorMsg,
                maxValue,
                minValue,
            }),
        };
    }
    preGetValidation() {
        const { value } = this.options;
        return this.getBasicValidation(value);
    }
    postGetValidation() {
        return this.getBasicValidation();
    }
    postValidateAndToast() {
        const validation = this.getValidation();
        if (validation.result === true) {
            return true;
        }
        this.showBubble(validation.errorMsg);
        return false;
    }
    emitManualStopEditEvent() {
        this.fireEvent('stopedit');
    }
    setInteractValue(v) {
        this.setControllerValue(v);
    }
    reset() {
        this.editor.setValue();
    }
    resize(width, height) {
        const restoreValue = this.editor.getValue();
        super.resize(width, height);
        this.editor.setValue(restoreValue);
    }
    getAdapter() {
        return this.editor;
    }
    focus() {
        this.editor.focus();
    }
    blur() {
        this.editor.blur();
    }
    setStyle(styles) {
        this.editor.element.addClass(setEditorStyle(styles));
    }
    setWaterMark(v) {
        this.editor.setWaterMark(v);
    }
    getText() {
        return this.editor.getValue();
    }
};
ParameterNumber.xtype = 'report.main.parameter_container.controller.number';
ParameterNumber = __decorate([
    shortcut()
], ParameterNumber);
export { ParameterNumber };
