// @engine/table/bin/modules/cell/cell_controllers/checkbox/cell_controller.checkbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { HEIGHT, ParameterCheckbox, WIDTH } from '@engine/controller';
import { shortcut } from '@engine/core';
import { AbstractParameterCellControllerWidget } from '../cell_controller.abstract';
let CellCheckBox = class CellCheckBox extends AbstractParameterCellControllerWidget {
    constructor() {
        super(...arguments);
        this.props = {
            width: WIDTH,
            height: HEIGHT,
            fontSize: 12,
            reportIndex: 0,
            position: {
                row: 0,
                column: 0,
            },
            xlisteners: [],
            baseCls: '',
        };
    }
    render() {
        const _a = this.options, { xlisteners, position, reportIndex, baseCls } = _a, opts = __rest(_a, ["xlisteners", "position", "reportIndex", "baseCls"]);
        return Object.assign(Object.assign({}, opts), { type: ParameterCheckbox.xtype, ref: (ref) => {
                this.checkbox = ref;
            }, listeners: xlisteners });
    }
    postValidateAndToast() {
        return this.checkbox.postValidateAndToast();
    }
    preGetValidation() {
        return this.checkbox.preGetValidation();
    }
    postGetValidation() {
        return this.checkbox.postGetValidation();
    }
    getValue() {
        return this.checkbox.getValue();
    }
    async setControllerValue(value) {
        await this.checkbox.setControllerValue(value);
    }
    setInteractValue(v) {
        this.checkbox.setInteractValue(v);
    }
    reset() {
        this.checkbox.reset();
    }
    setStyle() { }
};
CellCheckBox.xtype = 'report.main.table_cell.controller.checkbox';
CellCheckBox = __decorate([
    shortcut()
], CellCheckBox);
export { CellCheckBox };
