// @engine/table/bin/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { AxisSheet } from './modules/table';
export { tableCellProvider, tableCellTypeProvider, transformCellTypeByOpts } from './modules/cell/cell';
export { ImageCell, ImageTableCell } from './modules/cell/cell.image';
export { TextTableCell } from './modules/cell/cell.text';
export { AbstractCellControllerHoc } from './modules/cell/cell_controllers/cell_controller_hoc/cell_controller_hoc.abstract';
export { tableCellValueTransformerProvider } from './modules/cell/transformer/transformer';
export { NormalCellValueTransformer } from './modules/cell/transformer/transformer.normal';
export { tableClickHandlerProvider } from './modules/handler/handler';
export { AbstractCellHandler } from './modules/handler/handler.abstract';
export { WidgetActiveWithClickHandler } from './modules/handler/handler.active_click_controller';
export { WidgetActiveHandler } from './modules/handler/handler.active_controller';
export { NodeCellClickHandler } from './modules/handler/node/handler.node';
export { resizerProvider } from './modules/resizer/resizer';
export { AbstractResizer } from './modules/resizer/resizer.abstract';
export { FullPaperResizer } from './modules/resizer/resizer.full_paper';
export { FullTableResizer } from './modules/resizer/resizer.full_table';
export { FullTableLimitResizer } from './modules/resizer/resizer.full_table_limit';
export { HorizontalPaperResizer } from './modules/resizer/resizer.horizontal_paper';
export { HorizontalTableResizer } from './modules/resizer/resizer.horizontal_table';
export { HorizontalTableLimitResizer } from './modules/resizer/resizer.horizontal_table_limit';
export { NoneResizer } from './modules/resizer/resizer.none';
export { NoneTableResizer } from './modules/resizer/resizer.none_table';
export { AxisSheet } from './modules/table';
export { CellControllerManager } from './modules/table.cell_controller_manager';
export { CellErrorTipManager } from './modules/table.cell_error_tip_manager';
export { CellCacheManager } from './modules/table.cell_value_manager';
export { getImgLayout, transformCellCv } from './modules/table.service';
/**
 * 创建表格入口
 * @param options 表格参数
 * @param options.container 表格渲染区域，可选dom节点以及选择器
 * @returns 表格实例
 */
export function createSheet(options) {
    const { container } = options, args = __rest(options, ["container"]);
    return BI.createWidget(Object.assign(Object.assign({}, args), { type: AxisSheet.xtype, element: container }));
}
