// @engine/utils/bin/modules/keycode/keycode.js
export const KEYCODE_TYPES = {
    modifier: [
        'Alt',
        'AltGraph',
        'CapsLock',
        'Control',
        'Fn',
        'FnLock',
        'Meta',
        'NumLock',
        'ScrollLock',
        'Shift',
        'Symbol',
        'SymbolLock', //  The Symbol Lock key.
    ],
    legacyModifier: [
        'Hyper',
        'Super', //  The Super key.
    ],
    whiteSpace: [
        'Enter',
        'Tab', //  The Horizontal Tabulation Tab key. The space or spacebar key is encoded as ' '.
    ],
    navigation: [
        'ArrowDown',
        'ArrowLeft',
        'ArrowRight',
        'ArrowUp',
        'End',
        'Home',
        'PageDown',
        'PageUp', //  The Page Up key, to scroll up or display previous page of content.
    ],
    editing: [
        'Backspace',
        'Clear',
        'Copy',
        'CrSel',
        'Cut',
        'Delete',
        'EraseEof',
        'ExSel',
        'Insert',
        'Paste',
        'Redo',
        'Undo', //  Undo the last action. (APPCOMMAND_UNDO)
    ], ui: [
        'Accept',
        'Again',
        'Attn',
        'Cancel',
        'ContextMenu',
        'Escape',
        'Execute',
        'Find',
        'Help',
        'Pause',
        'Play',
        'Props',
        'Select',
        'ZoomIn',
        'ZoomOut', //  The ZoomOut key. (KEYCODE_ZOOM_OUT)
    ], device: [
        'BrightnessDown',
        'BrightnessUp',
        'Eject',
        'LogOff',
        'Power',
        'PowerOff',
        'PrintScreen',
        'Hibernate',
        'Standby',
        'WakeUp', //  The WakeUp key. (KEYCODE_WAKEUP)
    ], imeCompositionKeys: [
        'AllCandidates',
        'Alphanumeric',
        'CodeInput',
        'Compose',
        'Convert',
        'Dead',
        'FinalMode',
        'GroupFirst',
        'GroupLast',
        'GroupNext',
        'GroupPrevious',
        'ModeChange',
        'NextCandidate',
        'NonConvert',
        'PreviousCandidate',
        'Process',
        'SingleCandidate', //  The Single Candidate function key.
    ], koreanSpecific: [
        'HangulMode',
        'HanjaMode',
        'JunjaMode', //  The Junja (Korean characters) Mode key.
    ], japaneseSpecific: [
        'Eisu',
        'Hankaku',
        'Hiragana',
        'HiraganaKatakana',
        'KanaMode',
        'KanjiMode',
        'Katakana',
        'Romaji',
        'Zenkaku',
        'ZenkakuHankaku', //  The Zenkaku/Hankaku (full-width/half-width) toggle key. (KEYCODE_ZENKAKU_HANKAKU)
    ],
    //   The exact number of these general purpose function keys varies on different platforms, and only the first few are defined explicitly here. Additional function key names are implicitly defined by incrementing the base-10 index at the end of the function key name. Thus, 'F24' and 'Soft8' are all valid key values.
    commonFunction: [
        'F1',
        'F2',
        'F3',
        'F4',
        'F5',
        'F6',
        'F7',
        'F8',
        'F9',
        'F10',
        'F11',
        'F12',
        'Soft1',
        'Soft2',
        'Soft3',
        'Soft4', //  General purpose virtual function key, as index 4.
    ],
    //   These are extra keys found on 'multimedia' keyboards.
    multimedia: [
        'ChannelDown',
        'ChannelUp',
        'Close',
        'MailForward',
        'MailReply',
        'MailSend',
        'MediaClose',
        'MediaFastForward',
        'MediaPause',
        'MediaPlay',
        'MediaPlayPause',
        'MediaRecord',
        'MediaRewind',
        'MediaStop',
        'MediaTrackNext',
        'MediaTrackPrevious',
        'New',
        'Open',
        'Print',
        'Save',
        'SpellCheck', //  Spellcheck the current document or selection. (APPCOMMAND_SPELL_CHECK)
    ],
    //   The normal 0 ... 9 numpad keys are encoded as '0' ... '9', but some multimedia keypads have buttons numbered from 1 ... 12. In these instances, the 10 key is often labeled 10 /0. Note: The 10 or 10 /0 key MUST be assigned a key value of '0'.
    multimediaNumpad: [
        'Key11',
        'Key12', //  The 12 key found on media numpads that	have buttons from 1 ... 12.
    ],
    //   Multimedia keys related to audio.
    audio: [
        'AudioBalanceLeft',
        'AudioBalanceRight',
        'AudioBassBoostDown',
        'AudioBassBoostToggle',
        'AudioBassBoostUp',
        'AudioFaderFront',
        'AudioFaderRear',
        'AudioSurroundModeNext',
        'AudioTrebleDown',
        'AudioTrebleUp',
        'AudioVolumeDown',
        'AudioVolumeUp',
        'AudioVolumeMute',
        'MicrophoneToggle',
        'MicrophoneVolumeDown',
        'MicrophoneVolumeUp',
        'MicrophoneVolumeMute', //  Mute the microphone. (APPCOMMAND_MICROPHONE_VOLUME_MUTE, KEYCODE_MUTE)
    ],
    //   Multimedia keys related to speech recognition.
    speech: [
        'SpeechCorrectionList',
        'SpeechInputToggle', //  Toggle between dictation mode and command/control mode. (APPCOMMAND_DICTATE_OR_COMMAND_CONTROL_TOGGLE)
    ],
    //   The Application Keys are special keys that are assigned to launch a particular application. Additional application key names can be defined by concatenating 'Launch' with the name of the application.
    application: [
        'LaunchApplication1',
        'LaunchApplication2',
        'LaunchCalendar',
        'LaunchContacts',
        'LaunchMail',
        'LaunchMediaPlayer',
        'LaunchMusicPlayer',
        'LaunchPhone',
        'LaunchScreenSaver',
        'LaunchSpreadsheet',
        'LaunchWebBrowser',
        'LaunchWebCam',
        'LaunchWordProcessor', //  The 'Word Processor' key.
    ],
    browser: [
        'BrowserBack',
        'BrowserFavorites',
        'BrowserForward',
        'BrowserHome',
        'BrowserRefresh',
        'BrowserSearch',
        'BrowserStop', //  Stop loading the current page or content. (APPCOMMAND_BROWSER_STOP)
    ],
    mobilePhone: [
        'AppSwitch',
        'Call',
        'Camera',
        'CameraFocus',
        'EndCall',
        'GoBack',
        'GoHome',
        'HeadsetHook',
        'LastNumberRedial',
        'Notification',
        'MannerMode',
        'VoiceDial', //  The Voice Dial key.
    ], tv: [
        'TV',
        'TV3DMode',
        'TVAntennaCable',
        'TVAudioDescription',
        'TVAudioDescriptionMixDown',
        'TVAudioDescriptionMixUp',
        'TVContentsMenu',
        'TVDataService',
        'TVInput',
        'TVInputComponent1',
        'TVInputComponent2',
        'TVInputComposite1',
        'TVInputComposite2',
        'TVInputHDMI1',
        'TVInputHDMI2',
        'TVInputHDMI3',
        'TVInputHDMI4',
        'TVInputVGA1',
        'TVMediaContext',
        'TVNetwork',
        'TVNumberEntry',
        'TVPower',
        'TVRadioService',
        'TVSatellite',
        'TVSatelliteBS',
        'TVSatelliteCS',
        'TVSatelliteToggle',
        'TVTerrestrialAnalog',
        'TVTerrestrialDigital',
        'TVTimer', //  Timer programming. (KEYCODE_TV_TIMER_PROGRAMMING)
    ],
    //   The key attribute values for media controllers (e.g. remote controls for television, audio systems, and set-top boxes) are derived in part from the consumer electronics technical specifications:
    mediaControls: [
        'AVRInput',
        'AVRPower',
        'ColorF0Red',
        'ColorF1Green',
        'ColorF2Yellow',
        'ColorF3Blue',
        'ColorF4Grey',
        'ColorF5Brown',
        'ClosedCaptionToggle',
        'Dimmer',
        'DisplaySwap',
        'DVR',
        'Exit',
        'FavoriteClear0',
        'FavoriteClear1',
        'FavoriteClear2',
        'FavoriteClear3',
        'FavoriteRecall0',
        'FavoriteRecall1',
        'FavoriteRecall2',
        'FavoriteRecall3',
        'FavoriteStore0',
        'FavoriteStore1',
        'FavoriteStore2',
        'FavoriteStore3',
        'Guide',
        'GuideNextDay',
        'GuidePreviousDay',
        'Info',
        'InstantReplay',
        'Link',
        'ListProgram',
        'LiveContent',
        'Lock',
        'MediaApps',
        'MediaAudioTrack',
        'MediaLast',
        'MediaSkipBackward',
        'MediaSkipForward',
        'MediaStepBackward',
        'MediaStepForward',
        'MediaTopMenu',
        'NavigateIn',
        'NavigateNext',
        'NavigateOut',
        'NavigatePrevious',
        'NextFavoriteChannel',
        'NextUserProfile',
        'OnDemand',
        'Pairing',
        'PinPDown',
        'PinPMove',
        'PinPToggle',
        'PinPUp',
        'PlaySpeedDown',
        'PlaySpeedReset',
        'PlaySpeedUp',
        'RandomToggle',
        'RcLowBattery',
        'RecordSpeedNext',
        'RfBypass',
        'ScanChannelsToggle',
        'ScreenModeNext',
        'Settings',
        'SplitScreenToggle',
        'STBInput',
        'STBPower',
        'Subtitle',
        'Teletext',
        'VideoModeNext',
        'Wink',
        'ZoomToggle', //  Toggle between full-screen and scaled content, or alter magnification level. (VK_ZOOM, KEYCODE_TV_ZOOM_MODE)
    ],
};
