// @engine/core/bin/modules/components/report_html/html.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { htmlDecode, shortcut } from '../../core.service';
import './html.less';
let ReportHtml = class ReportHtml extends BI.Widget {
    constructor() {
        super(...arguments);
        this.props = {
            baseCls: 'report-html',
            text: '',
            textAlign: 'left',
            lineHeight: null,
            whiteSpace: 'normal',
        };
    }
    render() {
        const _a = this.options, { text } = _a, args = __rest(_a, ["text"]);
        return Object.assign(Object.assign({}, args), { type: BI.Html.xtype, ref: (_ref) => {
                this.htmlContainer = _ref;
            }, text: htmlDecode(text) });
    }
    setText(text) {
        this.htmlContainer.setText(htmlDecode(text));
    }
};
ReportHtml.xtype = 'report.components.html';
ReportHtml = __decorate([
    shortcut()
], ReportHtml);
export { ReportHtml };
