// @engine/chart/bin/modules/chart.js
import { ReportChartAlpha } from './adapter/chart_alpha';
import { ReportChartBeta } from './adapter/chart_beta';
import { ReportChartDelta } from './adapter/chart_delta';
import { ReportChartGamma } from './adapter/chart_gamma';
export class ReportChartProvider {
    constructor() {
        /**
         * 适配类集合，顺序从新到旧
         */
        this.list = [
            ReportChartDelta,
            ReportChartGamma,
            ReportChartBeta,
            ReportChartAlpha,
        ];
        this.chart = ReportChartAlpha;
    }
    initChart() {
        if (this.chart)
            return;
        const { version, buildTime } = VanCharts;
        const chart = this.list.find(reportChart => version === '1.0.0'
            ? new Date(buildTime) >= new Date(reportChart.BUILD_TIME)
            : version >= reportChart.VERSION);
        this.chart = chart || ReportChartBeta;
    }
    get xtype() {
        this.initChart();
        return this.chart.xtype;
    }
    /**
     * 隐藏全屏图表
     */
    hideFullScreenChart() {
        VanCharts.hideFullScreenChart();
    }
}
export const reportChartProvider = new ReportChartProvider();
