// @engine/link/bin/modules/hyper_link.handler.js
/**
 * Hyperlink点击处理，表格、图表通用
 */
export class HyperLinkClickHandler {
    constructor(attach) {
        this.hyperLinks = [];
        this.attach = () => {
            if (!this.popup) {
                attach({
                    type: BI.AbsoluteLayout.xtype,
                    items: [{
                            el: {
                                type: BI.DownListCombo.xtype,
                                belowMouse: true,
                                container: 'body',
                                trigger: '',
                                $testId: 'report-table-popup',
                                el: {
                                    type: BI.Layout.xtype,
                                },
                                ref: (_ref) => {
                                    this.popup = _ref;
                                },
                                listeners: [
                                    {
                                        eventName: BI.DownListCombo.EVENT_CHANGE,
                                        action: (v) => {
                                            this.hyperLinks[v] &&
                                                this.hyperLinks[v].excute();
                                        },
                                    },
                                ],
                            },
                            top: 0,
                            left: 0,
                        }],
                });
            }
        };
    }
    handle(e, _hyperLinks) {
        this.attach();
        const hyperLinks = _hyperLinks
            .filter(hyperLink => {
            if (!hyperLink) {
                return false;
            }
            if (hyperLink.silent) {
                hyperLink.excute();
                return false;
            }
            return true;
        });
        if (hyperLinks.length === 0) {
            return;
        }
        if (hyperLinks.length === 1) {
            hyperLinks[0].excute();
            return;
        }
        const hyperLinksItems = hyperLinks
            .map((hyperLink, index) => [{
                $testId: 'report-table-popup-hyper-link-item',
                text: hyperLink.getName(),
                value: index,
            }]);
        this.hyperLinks = hyperLinks;
        this.popup.populate(hyperLinksItems);
        this.popup.showView(e);
    }
}
